lib/libmuser.c

```diff
@@ -1148,89 +1180,70 @@ lm_ctx_create(lm_dev_info_t *dev_info)
+        lm_ctx->irqs.efds[i] = -1;
+    }
+    lm_ctx->irqs.err_efd = -1;
+    lm_ctx->irqs.req_efd = -1;
+    lm_ctx->irqs.type = IRQ_NONE;
+    lm_ctx->irqs.max_ivs = max_ivs;
 
+    // Set other context data.
+    lm_ctx->pvt = dev_info->pvt;
+    lm_ctx->log = dev_info->log;
+    lm_ctx->log_lvl = dev_info->log_lvl;
+    lm_ctx->reset = dev_info->reset;
+
+    // Bounce the provided pci_info into the context.
     memcpy(&lm_ctx->pci_info, &dev_info->pci_info, sizeof(lm_pci_info_t));
 
-    if (!memcmp(&lm_ctx->pci_info.reg_info[LM_DEV_CFG_REG_IDX], &zero_reg_info,
-                sizeof zero_reg_info)) {
-        static const lm_reg_info_t default_cfg_reg_info = {
-            .flags = LM_REG_FLAG_RW,
-            .size = PCI_CFG_SPACE_SIZE,
-            .fn = noop_cb
-        };
-        lm_ctx->pci_info.reg_info[LM_DEV_CFG_REG_IDX] = default_cfg_reg_info;
+    // Setup the PCI config space for this context.
+    err = pci_config_setup(lm_ctx, dev_info);
+    if (err != 0) {
+        goto out;
     }
 
+    // Bounce info for the sparse mmap areas.
     err = copy_sparse_mmap_areas(lm_ctx->pci_info.reg_info,
                                  dev_info->pci_info.reg_info);
-    if (err)
+    if (err) {
         goto out;
+    }
 
+    // Attach to the muser control device.
     lm_ctx->fd = dev_attach(dev_info->uuid);
     if (lm_ctx->fd == -1) {
         err = errno;
         goto out;
     }
 
+    // Create the internal DMA controller.
     lm_ctx->dma = dma_controller_create(LM_DMA_REGIONS);
     if (lm_ctx->dma == NULL) {
         err = errno;
         goto out;
     }
 
-    lm_ctx->extended = dev_info->extended;
-    if (lm_ctx->extended) {
-        size = PCI_CFG_SPACE_EXP_SIZE;
-    } else {
-        size = PCI_CFG_SPACE_SIZE;
-    }
-    lm_ctx->pci_config_space = calloc(PCI_CFG_SPACE_EXP_SIZE, 1);
-    if (lm_ctx->pci_config_space == NULL) {
-        err = errno;
-        goto out;
-    }
-
-    init_pci_hdr(&lm_ctx->pci_config_space->hdr,
-                 &dev_info->pci_info.id,
-                 &dev_info->pci_info.cc,
-                 &dev_info->pci_info.ss,
-                 dev_info->pci_info.irq_count[LM_DEV_INTX_IRQ_IDX]);
-
-    for (i = 0; i < ARRAY_SIZE(lm_ctx->pci_config_space->hdr.bars); i++) {
-        if ((dev_info->pci_info.reg_info[i].flags & LM_REG_FLAG_MEM) == 0) {
-            lm_ctx->pci_config_space->hdr.bars[i].io.region_type |= 0x1;
-        }
-    }
-
-    lm_ctx->reset = dev_info->reset;
-    lm_ctx->pvt = dev_info->pvt;
-
-    for (i = 0; i < max_ivs; i++) {
-        lm_ctx->irqs.efds[i] = -1;
-    }
-    lm_ctx->irqs.err_efd = -1;
-    lm_ctx->irqs.req_efd = -1;
-    lm_ctx->irqs.type = IRQ_NONE;
-    lm_ctx->irqs.max_ivs = max_ivs;
-
-    lm_ctx->log = dev_info->log;
-    lm_ctx->log_lvl = dev_info->log_lvl;
-
-    err = lm_caps_init(lm_ctx, dev_info->caps, dev_info->nr_caps);
-
 out:
     if (err) {
         if (lm_ctx) {
+            dma_controller_destroy(lm_ctx, lm_ctx->dma);
```

Comment:
That should have been an individual commit.

Reply:
Probably. As per comment above, I made a trade-off to try and clean up the whole function in one go.